src/envs/cluttered_table.py

```diff
@@ -267,3 +242,122 @@ def _any_intersection(pose: Array, radius: float,
             if distance <= (radius + other_radius):
                 return True
         return False
+
+    @staticmethod
+    def _check_collisions(start_x: float,
+                          start_y: float,
+                          end_x: float,
+                          end_y: float,
+                          state: State,
+                          desired_can: Optional[Object] = None) -> None:
+        """Handle collision checking.
+
+        We'll just threshold the angle between the grasp approach vector
+        and the vector between the desired_can and any other can. Doing
+        an actually correct geometric computation would involve the
+        radii somehow, but we don't really care about this.
+        """
+        vec1 = np.array([end_x - start_x, end_y - start_y])
+        colliding_can = None
+        colliding_can_max_dist = float("-inf")
+        for can in state:
+            if can == desired_can:
+                continue
+            this_x = state.get(can, "pose_x")
+            this_y = state.get(can, "pose_y")
+            vec2 = np.array([end_x - this_x, end_y - this_y])
+            angle = np.arccos(
+                np.clip(
+                    vec1.dot(vec2) / (
+                        np.linalg.norm(vec1) *  # type: ignore
+                        np.linalg.norm(vec2)),  # type: ignore
+                    -1.0,
+                    1.0))
+            if abs(angle) < CFG.cluttered_table_collision_angle_thresh:
+                dist = np.linalg.norm(vec2)  # type: ignore
+                if dist > colliding_can_max_dist:
+                    colliding_can_max_dist = float(dist)
+                    colliding_can = can
+        if colliding_can is not None:
+            raise EnvironmentFailure("collision", {colliding_can})
+
+
+class ClutteredTablePlaceEnv(ClutteredTableEnv):
+    """Toy cluttered table domain (place version).
+
+    This version places grasped cans instead of dumping them. As an
+    additional challenge, the action space is restricted so that actions
+    can only begin from within a 0.2 x 0.2 corner
```

Comment:
won't this make the env impossible?

Reply:
maybe would be good to clarify what are the implications of this change planning-wise (it's not super clear to me)

Reply:
I probably wasn't clear enough here. The goal is to learn to pick up colliding cans between the robot (in the 0.2x0.2 corner) and the desired can, and then place them somewhere where they are not in the way. I updated the comment in the file to say so! (The actual learning about the failure is something Tom and I have yet to work on and is the motivation for this environment).